### PR TITLE
Fix for APIMANAGER-6048

### PR DESCRIPTION
**Pull request by uvindra**
### Proposed changes in this pull request
Fix for APIMANAGER-6048. This involves moving the sample entitlement mediator used by the XACML usage example[1] into the carbon-apimgt code base to remove the burden of updating dependency versions in new product releases 

[1] https://docs.wso2.com/display/AM210/Enabling+Role-Based+Access+Control+Using+XACML

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+public class APIEntitlementCallbackHandler extends EntitlementCallbackHandler {
+    private static final Log log = LogFactory.getLog(APIEntitlementCallbackHandler.class);
+
+    public String getUserName(MessageContext synCtx) {
+        AuthenticationContext authContext = APISecurityUtils.getAuthenticationContext(synCtx);
+        String userName = null;
+        if (authContext != null) {
+            userName = authContext.getUsername();
+
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: extra line can be removed.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        }
+        log.debug("UserName ---"+userName);
+        return userName;
+    }
+
+    public String findServiceName(MessageContext synCtx) {
+        String path = ((String) synCtx.getProperty(RESTConstants.REST_SUB_REQUEST_PATH));
+        log.debug("SERVICE - REST SUB REQUEST ---"+path);
+        return path;
+
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: extra line

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public String findAction(MessageContext synCtx) {
+        log.debug("Operation Name ---"+(String)((Axis2MessageContext) synCtx).getAxis2MessageContext().getProperty(org.apache.axis2.Constants.Configuration.HTTP_METHOD));
+        return (String)((Axis2MessageContext) synCtx).getAxis2MessageContext().getProperty(org.apache.axis2.Constants.Configuration.HTTP_METHOD);
+
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: extra line

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java (hunk 2 above)
Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: formatting issue. spaces required.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    public String findServiceName(MessageContext synCtx) {
+        String path = ((String) synCtx.getProperty(RESTConstants.REST_SUB_REQUEST_PATH));
+        log.debug("SERVICE - REST SUB REQUEST ---"+path);
+        return path;
+
+    }
+
+    public String findAction(MessageContext synCtx) {
+        log.debug("Operation Name ---"+(String)((Axis2MessageContext) synCtx).getAxis2MessageContext().getProperty(org.apache.axis2.Constants.Configuration.HTTP_METHOD));
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: formatting issue. spaces required.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            userName = authContext.getUsername();
+
+        }
+        log.debug("UserName ---"+userName);
+        return userName;
+    }
+
+    public String findServiceName(MessageContext synCtx) {
+        String path = ((String) synCtx.getProperty(RESTConstants.REST_SUB_REQUEST_PATH));
+        log.debug("SERVICE - REST SUB REQUEST ---"+path);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: we need to check is debug enable

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java (hunk 4 above)
Comment (harsha89) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: check debug is enabled

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java (hunk 5 above)
Comment (uvindra) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: This is not going to be used at runtime by the actual product, its just a sample

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        log.debug("UserName ---" + userName);
+        return userName;
+    }
+
+    public String findServiceName(MessageContext synCtx) {
+        String path = ((String) synCtx.getProperty(RESTConstants.REST_SUB_REQUEST_PATH));
+        log.debug("SERVICE - REST SUB REQUEST ---" + path);
+        return path;
+
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: extra line

Comment (uvindra) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/sample/handlers/entitlement/APIEntitlementCallbackHandler.java]: removed

**Pull request merged**